File: Algorithms/152.MaximumProductSubarray.java
public class Solution {
    public int maxProduct(int[] A) {
        int n = A.length;
        int max = A[0], min = A[0], result = A[0];
        for (int i = 1; i < n; i++) {
            int mx = max, mn = min;
            max = Math.max(Math.max(A[i], mx * A[i]), mn * A[i]);
            min = Math.min(Math.min(A[i], mx * A[i]), mn * A[i]);
            result = Math.max(result, max);
        }
        return result;
    }
}

File: Algorithms/152.MaximumProductSubarray.cpp
class Solution {
public:
    int maxProduct(vector<int>& nums) {
        int max_product = nums[0], min_product = nums[0], result = nums[0];
        for (int i = 1; i < nums.size(); i++) {
            int mx = max_product, mn = min_product;
            max_product = max(max(nums[i], mx * nums[i]), mn * nums[i]);
            min_product = min(min(nums[i], mx * nums[i]), mn * nums[i]);
            result = max(result, max_product);
        }
        return result;
    }
};
